src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Completion/RazorDirectiveAttributeCompletionSource.cs

```diff
@@ -0,0 +1,257 @@
+                // we'll need to dismiss the legacy completion session to ensure we don't get two completion lists.
+                var activeSessions = _completionBroker.GetSessions(session.TextView);
+                foreach (var activeSession in activeSessions)
+                {
+                    if (activeSession.Properties.ContainsProperty(nameof(IAsyncCompletionSession)))
+                    {
+                        continue;
+                    }
+
+                    // Legacy completion is also active, we need to dismiss it.
+
+                    _ = Task.Factory.StartNew(
+                        () => activeSession.Dismiss(),
+                        CancellationToken.None,
+                        TaskCreationOptions.None,
+                        _foregroundDispatcher.ForegroundScheduler);
+                }
+
+                var completionItems = new List<CompletionItem>();
+                var completionItemKinds = new HashSet<RazorCompletionItemKind>();
+                foreach (var razorCompletionItem in razorCompletionItems)
+                {
+                    if (razorCompletionItem.Kind != RazorCompletionItemKind.DirectiveAttribute)
+                    {
+                        // Don't support any other types of completion kinds other than directive attributes.
+                        continue;
+                    }
+
+                    var completionItem = new CompletionItem(
+                        displayText: razorCompletionItem.DisplayText,
+                        filterText: razorCompletionItem.DisplayText,
+                        insertText: razorCompletionItem.InsertText,
+                        source: this,
+                        icon: DirectiveAttributeImageGlyph,
+                        filters: DirectiveAttributeCompletionFilters,
+                        suffix: string.Empty,
+                        sortText: razorCompletionItem.DisplayText,
+                        attributeIcons: ImmutableArray<ImageElement>.Empty);
+                    completionItems.Add(completionItem);
+                    completionItemKinds.Add(razorCompletionItem.Kind);
+                }
+
+                session.Properties.SetCompletionItemKinds(completionItemKinds);
+                var orderedCompletionItems = completionItems.OrderBy(item => item.DisplayText);
+                var context = new CompletionContext(orderedCompletionItems.ToImmutableArray());
+                return context;
+            }
+            catch (OperationCanceledException)
+            {
+                return CompletionContext.Empty;
+            }
+        }
+
+        public Task<object> GetDescriptionAsync(IAsyncCompletionSession session, CompletionItem item, CancellationToken token)
+        {
+            return Task.FromResult<object>(string.Empty);
+        }
+
+        public CompletionStartData InitializeCompletion(CompletionTrigger trigger, SnapshotPoint triggerLocation, CancellationToken token)
```

Comment:
This is the entry point of the completion engine. This will get called on every keypress in a Razor document, it's goal is to return an applicable span or a "does not participate result".

If this returns an applicable span the `GetCompletionsAsync` method above is called which indirectly calls our completion facts service. The completion items that are returned are then cross-referenced with the applicable-to-span to functionally impact sorting/filtering/committing etc. of items.